R/firstlast.R

```diff
@@ -0,0 +1,247 @@
+
+    if (length(by) == 1)
+        by <- rep(by, n.dim)
+
+    indices <- lapply(seq_len(n.dim), function(i) {
+        if (i > length(keep) || is.na(keep[i]))
+            return(seq_len(dim.x[i]))
+        nms <- if (is.array(x)) dimnames(x)[[i]] else names(x)
+        date.times <- parseDateTime(nms, by[i], i, n.dim)
+
+        period.integers <- periodIntegers(date.times, by[i], calendar, is.first)
+        ind <- if (is.first)
+            which(period.integers <= min(period.integers) + keep[i] - 1)
+        else
+            which(period.integers >= max(period.integers) - keep[i] + 1)
+        ind[order(date.times[ind])] # return indices in ascending date order
+    })
+    do.call(`[`, c(list(x), indices, drop = FALSE))
+}
+
+#' @importFrom flipTime AsDateTime
+parseDateTime <- function(date.time.strings, by, dimension.index,
+                          n.dimensions)
+{
+    common.msg <- if (n.dimensions == 1)
+        paste0("The duration '", by,
+               "' cannot be applied as the input data is ")
+    else if (n.dimensions == 2)
+    {
+        dimension.label <- if (dimension.index == 1) "rows" else "columns"
+        paste0("The duration '", by,
+               "' cannot be applied as the ", dimension.label,
+               " in the input data are ")
+    }
+    else
+        paste0("The duration '", by,
+               "' cannot be applied as dimension ", dimension.index,
+               " in the input data is ")
+
+    if (is.null(date.time.strings))
+        stop(common.msg, "not labeled with dates.")
+    date.times <- AsDateTime(date.time.strings, on.parse.failure = "")
+    if (any(is.na(date.times)))
+        stop(common.msg, "labeled with invalid date(s).")
+    date.times
+}
+
+# Represent periods in date.times as integers, where the smallest integer
+# corresponds to the earliest period and vice versa. The parameter 'from.start'
+# indicates whether to count periods from the first date as opposed to the last
+# date when calendar is FALSE.
+periodIntegers <- function(date.times, by, calendar, from.start)
+{
+    if (calendar)
+    {
+        start.date <- structure(0, class = c("POSIXct", "POSIXt")) # unix epoch
+        if (by == "week")
+            start.date <- start.date + 3 * 24 * 60 *60 # add 3 days to unix epoch
+        intervalLength(start.date, date.times, by)
+    }
+    else if (from.start)
+        intervalLength(min(date.times), date.times, by)
+    else
+        # negation causes latest date to have the greatest integer
+        -intervalLength(date.times, max(date.times), by)
+}
+
+# The number of time periods in units of 'by' between the start and end dates
+# as a integer (rounded down)
+#' @importFrom lubridate interval time_length
+intervalLength <- function(start.date.time, end.date.time, by)
+{
+    if (by == "quarter")
+        floor(intervalLength(start.date.time, end.date.time, "month") / 3)
+    else
+        floor(time_length(interval(start.date.time, end.date.time), by))
+}
```

Comment:
Could unit test these internal functions, `intervalLength` and `periodIntegers` and `parseDateTime`?

Reply:
Done in 08281c682f471a7ffa02a710253dd431017d29dc